src/GraphQL/Internal/Syntax/Parser.hs

```diff
@@ -51,7 +52,7 @@ operationDefinition =
   <?> "operationDefinition error!"
 
 node :: Parser AST.Node
-node = AST.Node <$> AST.nameParser
+node = AST.Node <$> optempty nameParser
```

Comment:
Why did you decide to go for `optempty` rather than `optional`?

My natural approach for modelling anonymous nodes would be for them to have a name of `Nothing`, rather than a name of `Name ""`.

Reply:
I didn't because it was a deeper change. I would have to change the definition of
```
class HasName a where
  getName :: a -> Name
```
to
```
class HasName a where
  getName :: a -> Maybe Name
```
and then handle the Maybe everywhere that `getName` is used. Is this a direction you'd like to go?

Reply:
Hmm. Changing the typeclass seems wrong too, since names are only optional for nodes (IIUC). 

I guess if we went this direction, we'd drop the typeclass for nodes and re-instate `getNodeName :: Node -> Maybe Name`.

Reply:
Well, that's amusing. I suppose that's why Node was NOT an instance of HasName to begin with.

It would have been much simpler to change:
```
getNodeName :: Node -> Name
getNodeName (Node name _ _ _) = name
```
to
```
getNodeName :: Node -> Maybe Name
getNodeName (Node name _ _ _) = name
```
with
```
data Node = Node (Maybe Name) [VariableDefinition] [Directive] SelectionSet
             deriving (Eq,Show)
```
The only way to used `getName` of `HasName` would be to use it poorly by returning an empty `Name` when when encounter a `Node` with `Nothing` as `Name`.

Reply:
> I suppose that's why Node was NOT an instance of HasName to begin with.

Could well have been. To be honest, I can't remember the details. 

> It would have been much simpler…

Yeah, it would have. I'm sorry I didn't think of this at the time of #134.

Reply:
Which would you prefer to do?

1. Go back to the original `getNodeName` and have it return `Maybe Name` with the corresponding change in `Node`. Cleaner with `Maybe`, but ugly and non-unified get function.
2. Keep instance of `HasName` for `Node`, change `Node` to 
```
data Node = Node (Maybe Name) [VariableDefinition] [Directive] SelectionSet
             deriving (Eq,Show)
```
AND have the `getName` return an empty `Name` when we encounter `Nothing` instead of `Just Name "somename"`
3. Something I haven't considered...?

Reply:
Please:
- change `Node` to be `Node (Maybe Name) [VariableDefinition] [Directive] SelectionSet`
- go back to original `getNodeName`, except returning `Maybe Name`
- drop the `HasName` instance for `Node`

I'm sorry for the confusion here. I did a bit of digging and it looks like the spec was changed to make names for `Node` optional while we were in the middle of implementing graphql-api.

Reply:
Will do. Tomorrow :)

Reply:
Turns out to not be that simple. There is some interaction between `Name` values and the new `Maybe Name` values in Validation.hs

At line 140 of Validation.hs `visitedFrags` which contains a `Set` of `Name` and `usedFrags` which contains a `Set` of `Maybe Name` are combined. I'm not yet sure how to handle this.